### PR TITLE
Feature/add document to issues and projects

### PR DESCRIPTION
**Pull request by eman-ijaz7vals**

### CODE REVIEW

**Comment by eman-ijaz7vals** [comment]
Cycle -0 reviewed

File: app/assets/javascripts/documents.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (abdullah-siddiqui-git) [inline app/assets/javascripts/documents.coffee]: Can delete this file as it seems useless.

File: app/assets/javascripts/projects.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+
+$( document ).ready(function() {
+  // Menu Toggle Script 
+  $("#menu-toggle").click(function(e) {
+      e.preventDefault();
+      $("#wrapper").toggleClass("toggled");
+  });
+
+  $("#myInput").on("keyup", function() {
```
Comment (abdullah-siddiqui-git) [inline app/assets/javascripts/projects.js]: Can give the input field a more meaningful id according to the context in which it is used.
E.g. #myInput -> #projects_filter_input_field

File: app/assets/javascripts/projects.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    var value = $(this).val().toLowerCase();
+    $("#myList tr").filter(function() {
```
Comment (abdullah-siddiqui-git) [inline app/assets/javascripts/projects.js]: Can change #myList to some more meaningful id.

File: app/assets/stylesheets/application.scss [hunk 4]
```diff
@@ -13,3 +13,21 @@
  *= require_tree .
  *= require_self
  */
+
+/* CSS for sidebar */
+
+/*!
```
Comment (abdullah-siddiqui-git) [inline app/assets/stylesheets/application.scss]: Can remove useless comments.

File: app/assets/stylesheets/custom.css.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+//@import "bootstrap-sprockets";
```
Comment (abdullah-siddiqui-git) [inline app/assets/stylesheets/custom.css.scss]: Can delete this if its useless.

File: app/controllers/comments_controller.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    @comment = Comment.new
+    respond_to do |format|
+      format.html
+      format.js
+    end
+  end
+
+  def create
+    @comment = @project.comments.build(comment_params)
+    @comment.company_id = 1
```
Comment (abdullah-siddiqui-git) [inline app/controllers/comments_controller.rb]: Why static company?
Shouldn't it be?
@comment.company_id = current_tenant.id

File: app/controllers/comments_controller.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  def update
+    if @comment.update(comment_params)
+      respond_to do |format|
+        format.js
+      end
+      flash[:notice] = t('comments.update.updated')
+    else
+      # how to show the errors
+      render 'edit'
+      flash[:notice] = t('comments.update.not_updated')
```
Comment (abdullah-siddiqui-git) [inline app/controllers/comments_controller.rb]: Why flashing out message after rendering the page?

File: app/controllers/comments_controller.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    end
+  end
+
+  def destroy
+    if @comment.destroy
+      flash[:notice] = t('comments.destroy.destroyed')
```
Comment (abdullah-siddiqui-git) [inline app/controllers/comments_controller.rb]: What if '.destroy' fails in the JS request and code flow jumps into else part? Will it render "destroy.html.erb" in the response? Moreover, it is encouraged to use only one responds_to block at the end of the method.

File: app/controllers/documents_controller.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    end
+  end
+
+  def new
+    @document = Document.new
+  end
+
+  def create
+    @document = Document.new(document_params)
+    @document.company_id = 1
```
Comment (abdullah-siddiqui-git) [inline app/controllers/documents_controller.rb]: Change to 
@document.company_id = current_tenant.id

Or 

current_tenant.documents.new(document_params)

File: app/controllers/documents_controller.rb [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    @document.path = @document.document.url
+    if params[:issue_id]
```
Comment (abdullah-siddiqui-git) [inline app/controllers/documents_controller.rb]: Can use params[:issue_id].nil?

File: app/controllers/documents_controller.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+  def destroy
+    @document = Document.find(params[:id])
+    if params[:issue_id]
+      @path = issue_path(params[:issue_id])
+    else
+      @path = project_path(params[:project_id])
+    end
+    if @document.destroy
+      redirect_to @path, delete_document: t('.Successfully deleted document!')
```
Comment (abdullah-siddiqui-git) [inline app/controllers/documents_controller.rb]: delete_document: t('.Successfully deleted document!') passes message in query string. We can use flash messages instead.

File: app/controllers/issues_controller.rb [hunk 12]
```diff
@@ -51,6 +50,7 @@ def edit
   # PUT /issues/:id
   def update
     if @issue.update(issue_params)
+      redirect_to @issue, update_issue: t('.Issue Updated successfully!')
```
Comment (abdullah-siddiqui-git) [inline app/controllers/issues_controller.rb]: Why redirect here, if we have respond_to block ahead in the code?

File: app/controllers/projects_controller.rb [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  def new
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def create
+    @project.company_id = 1
```
Comment (abdullah-siddiqui-git) [inline app/controllers/projects_controller.rb]: @project.company_id = current_tenant.id

File: app/controllers/teams_controller.rb [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    @all_teams = Team.all
+  end
+
+  def new
+    @team = Team.new
+  end
+
+  def show
+    @team = Team.find(params[:id])
+    @team_memberships = TeamMembership.where('team_id = ? ', @team.id)
```
Comment (abdullah-siddiqui-git) [inline app/controllers/teams_controller.rb]: Can change syntax to:
@team_memberships = TeamMembership.where(team_id: @team.id)

File: app/controllers/teams_controller.rb [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    already_present_member_ids = @team_memberships.pluck(:user_id)
+    @all_members = User.where.not(id: already_present_member_ids).select(:id, :first_name)
+  end
+
+  def create
+    @team = Team.new(team_params)
+    @team.company_id = 1 ## need to be changed
```
Comment (abdullah-siddiqui-git) [inline app/controllers/teams_controller.rb]: @team.company_id = current_tenant.id

File: app/controllers/teams_controller.rb [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+    if @team.save
+      @user_id = current_user.id
+      @team_id = Team.last.id
+      TeamMembership.create!(is_team_admin: true, is_approved: false, team_id: @team_id, user_id: @user_id) ## person who
```
Comment (abdullah-siddiqui-git) [inline app/controllers/teams_controller.rb]: Could make it a transaction to ensure atomicity.

File: app/controllers/teams_controller.rb [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+  # change status of user "is_approved = true" on request to join team
+  def join_team
+    @is_approved = true
+    add_membership(@is_approved)
+  end
+
+  def add_membership(is_approved)
+    @user_id = params[:user_id]
+    @team_id = params[:team_id]
+    @is_admin = if params[:join_admin][:result] == '0'
```
Comment (abdullah-siddiqui-git) [inline app/controllers/teams_controller.rb]: @is_admin = !params[:join_admin][:result] == '0'

File: app/views/comments/_all_comments.html.erb [hunk 18]
```diff
@@ -0,0 +1,39 @@
+<% @comments.each do |comment| %>
+  <% next if comment.new_record? %>
+  <div id="this_comment_<%= comment.id %>" class="card v-card v-sheet theme--light elevation-2">
+    <div id="header" class="header">
+      <div  class="v-avatar avatar" style="height: 50px; width: 50px;"><img  src="https://images.unsplash.com/photo-1490894641324-cfac2f5cd077?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=100&q=70">
+      </div>
+      <span  class="displayName title">John Doe <%= comment.user_id %></span>
```
Comment (abdullah-siddiqui-git) [inline app/views/comments/_all_comments.html.erb]: Hard coded name.

File: app/views/comments/_single_comment.html.erb [hunk 19]
```diff
@@ -0,0 +1,38 @@
+
+<div id="this_comment_<%= comment.id %>" class="card v-card v-sheet theme--light elevation-2">
+  <div  class="header">
+    <div  class="v-avatar avatar" style="height: 50px; width: 50px;"><img  src="https://images.unsplash.com/photo-1490894641324-cfac2f5cd077?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=100&q=70">
+    </div>
+    <span  class="displayName title">John Doe <%= comment.user_id %></span>
```
Comment (abdullah-siddiqui-git) [inline app/views/comments/_single_comment.html.erb]: Hard coded name.

File: spec/helpers/comments_helper_spec.rb [hunk 20]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
+
```
Comment (abdullah-siddiqui-git) [inline spec/helpers/comments_helper_spec.rb]: Too many useless comments.

File: app/controllers/comments_controller.rb [hunk 21]
```diff
@@ -0,0 +1,70 @@
+# Controller for comments model. Comments are to be shown on projects and issues
+class CommentsController < ApplicationController
+  before_action :load_project, except: [:edit, :update, :destroy]
```
Comment (mujtaba7vals) [inline app/controllers/comments_controller.rb]: load_and_authorize resource

**Comment by abdullah-siddiqui-git** [comment]
Cycle - 1 review done.

File: app/controllers/documents_controller.rb [hunk 22]
```diff
@@ -0,0 +1,73 @@
+class DocumentsController < ApplicationController
+  def index
+    @issue_id = params[:issue_id]
+    @project_id = params[:project_id]
```
Comment (mujtaba7vals) [inline app/controllers/documents_controller.rb]: params can be used directly

File: app/controllers/documents_controller.rb [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+    if params[:issue_id] ## called via issue
```
Comment (mujtaba7vals) [inline app/controllers/documents_controller.rb]: .present?

File: app/controllers/documents_controller.rb [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+      @issue_documents = Issue.find(params[:issue_id]).documents
+    else ## called via projects
+      @project_documents = Project.find(params[:project_id]).documents
+
```
Comment (mujtaba7vals) [inline app/controllers/documents_controller.rb]: spacing

File: app/controllers/documents_controller.rb [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+      redirect_to @path, save_document: t('.save_document') # redirect to show page
+    else
+      render 'new', document_not_saved: t('.Document not saved!')
+    end
+  end
+
+  def destroy
+    @document = Document.find(params[:id])
+    if params[:issue_id]
+      @path = issue_path(params[:issue_id])
```
Comment (mujtaba7vals) [inline app/controllers/documents_controller.rb]: instance var not required

File: app/controllers/documents_controller.rb (hunk 11 above)
Comment (mujtaba7vals) [inline app/controllers/documents_controller.rb]: respond_to blocks needed

File: app/controllers/documents_controller.rb [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+    else
+      @path = project_path(params[:project_id])
+    end
+    if @document.destroy
+      redirect_to @path, delete_document: t('.Successfully deleted document!')
+    else
+      render 'index', document_not_delete: t('.Error deleting Document!')
```
Comment (mujtaba7vals) [inline app/controllers/documents_controller.rb]: use flash[:error]

File: app/controllers/issues_controller.rb [hunk 27]
```diff
@@ -33,9 +33,8 @@ def new
 
   # GET /issues/:id
   def show
-    respond_to do |format|
-      format.html
-    end
+    @issue = Issue.find(params[:id])
+    @document = Document.new
```
Comment (mujtaba7vals) [inline app/controllers/issues_controller.rb]: respond_to

File: app/controllers/teams_controller.rb [hunk 28]
```diff
@@ -0,0 +1,92 @@
+class TeamsController < ApplicationController
+  load_and_authorize_resource
+
+  def index
+    @all_teams = Team.all
```
Comment (mujtaba7vals) [inline app/controllers/teams_controller.rb]: not needed

File: app/controllers/teams_controller.rb (hunk 16 above)
Comment (mujtaba7vals) [inline app/controllers/teams_controller.rb]: rescue clause needed

Comment (mujtaba7vals) [inline app/controllers/teams_controller.rb]: use directly

File: app/controllers/teams_controller.rb [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+      redirect_to teams_path, team_updated: t('.Team Successfully updated!')
+    else
+      render 'edit'
+    end
+  end
+
+  def destroy
+    @team = Team.find(params[:id])
+    if @team.destroy
+      redirect_to teams_path, team_deleted: t('.Successfully deleted team!')
```
Comment (mujtaba7vals) [inline app/controllers/teams_controller.rb]: flash[:error] = t('.Successfully deleted team!')

File: app/controllers/teams_controller.rb [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+    else
+      redirect_to teams_path team_not_deleted: t('.Error deleting team!')
```
Comment (mujtaba7vals) [inline app/controllers/teams_controller.rb]: use a single redirect_to with respond_do block

File: app/controllers/teams_controller.rb [hunk 31]
```diff
@@ -0,0 +1,92 @@
+class TeamsController < ApplicationController
+  load_and_authorize_resource
+
+  def index
```
Comment (mujtaba7vals) [inline app/controllers/teams_controller.rb]: # GET  /index 
add restful comments on each action

File: app/controllers/teams_controller.rb [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def team_membership_params
+    params.require(:team_memberships).permit(:is_team_admin, :is_approved, :team_id, :user_id)
+  end
+
+  ## add member in team membership
+  def add_member
+    @is_approved = false
+    add_membership(@is_approved)
```
Comment (mujtaba7vals) [inline app/controllers/teams_controller.rb]: no need of extra instance variables

File: app/controllers/teams_controller.rb (hunk 17 above)
Comment (mujtaba7vals) [reply app/controllers/teams_controller.rb]: `if params[:join_admin][:result] == '0'` -> `params:[join][:result] != 0`

File: app/controllers/teams_controller.rb [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+  def add_membership(is_approved)
+    @user_id = params[:user_id]
+    @team_id = params[:team_id]
+    @is_admin = if params[:join_admin][:result] == '0'
+                  false
+                else
+                  true
+    end
+    @team_membership = TeamMembership.new(is_team_admin: @is_admin, is_approved: is_approved, team_id: @team_id, user_id: @user_id)
+    if @team_membership.save
```
Comment (mujtaba7vals) [inline app/controllers/teams_controller.rb]: handle failure case

File: app/models/issue.rb [hunk 34]
```diff
@@ -27,6 +27,8 @@ class Issue < ActiveRecord::Base
   has_many   :documents
 
   has_many   :comments, as: :commentable
+  has_many :documents, as: :documentable
```
Comment (mujtaba7vals) [inline app/models/issue.rb]: spacing

File: app/models/issue.rb [hunk 35]
```diff
@@ -27,6 +27,8 @@ class Issue < ActiveRecord::Base
...
+  has_many   :comments, as: :commentable, dependent: :destroy
```
Comment (mujtaba7vals) [inline app/models/issue.rb]: repetition

File: app/models/project.rb [hunk 36]
```diff
@@ -11,4 +16,9 @@ class Project < ActiveRecord::Base
                    source_type: 'Team'
   has_many :users, through: :project_memberships, source: :project_member,
                    source_type: 'User'
+
+  def visible?(user)
```
Comment (mujtaba7vals) [inline app/models/project.rb]: boolean method should true/false

File: app/views/comments/_all_comments.html.erb [hunk 37]
```diff
@@ -0,0 +1,39 @@
...
+      
+      <span  class="displayName caption"><%= time_ago_in_words(comment.created_at) %> ago</span>
+
+      <span  class="displayName caption"> <%= link_to "Edit", edit_comment_path(comment), remote: true %> </span>
+      <span  class="displayName caption"> <%= link_to 'Delete', comment_path(comment), method: :delete, data: { confirm: 'Are you sure?' }, remote: true %> </span>
+
+      <%= form_for comment, url: comment_path(comment), hidden: true, remote: true, html: {id: "comment_edit_form_"+comment.id.to_s, class: "comment_edit_form"} do |f| %>
```
Comment (mujtaba7vals) [inline app/views/comments/_all_comments.html.erb]: use `"#{}"`

File: app/views/comments/_editform.html.erb [hunk 38]
```diff
@@ -0,0 +1,23 @@
...
+        <li><%= message %></li>
+      <% end %>
+      </ul>
+    </div>
+  <% end %>
+
+  <div class="form-group">
+    <label for="message-text" class="col-form-label">Message:</label>
+
+    <%= f.text_area :content, id: "editcommenttext", class: 'form-control' %>
```
Comment (mujtaba7vals) [inline app/views/comments/_editform.html.erb]: change id name

File: app/views/documents/index.html.erb [hunk 39]
```diff
@@ -0,0 +1,66 @@
+<div class="container">
+
+  <% if (@issue_documents and @issue_documents.count == 0) or (@project_documents and @project_documents.count == 0)  %>
```
Comment (mujtaba7vals) [inline app/views/documents/index.html.erb]: use `&& ||`

File: app/views/documents/index.html.erb [hunk 40]
```diff
@@ -0,0 +1,66 @@
...
+    <h3>No documents to show</h3>
+
+  <% else %>
+  <h1>Documents List</h1>
```
Comment (mujtaba7vals) [inline app/views/documents/index.html.erb]: indent

File: app/views/documents/index.html.erb [hunk 41]
```diff
@@ -0,0 +1,66 @@
...
+
+  <table class="table table-hover" id="DocumentsTable">
+  <thead>
+    <tr>
+      <th scope="col">Document </th>
+      <th scope="col">Document Index</th>
+      <th scope="col">Name</th>
+      <th scope="col">Type</th>
+      <th scope="col">Download</th>
+
```
Comment (mujtaba7vals) [inline app/views/documents/index.html.erb]: spacing

File: app/views/documents/index.html.erb [hunk 42]
```diff
@@ -0,0 +1,66 @@
...
+      <th scope="col">Type</th>
+      <th scope="col">Download</th>
+    
+    </tr>
+  </thead>
+  <tbody>
+
+      <% if @issue_documents %>
+
+            <% @issue_documents.all.each_with_index do |document, index_no| %>
```
Comment (mujtaba7vals) [inline app/views/documents/index.html.erb]: all not needed

File: app/views/documents/index.html.erb [hunk 43]
```diff
@@ -0,0 +1,66 @@
...
+                  <tr>
+                    <% if document.document.url.include? 'jpeg' or document.document.url.include? 'png' or document.document.url.include? 'jpg'  %>
```
Comment (mujtaba7vals) [inline app/views/documents/index.html.erb]: use single .include? %w[jpeg png jpq]

File: app/views/documents/index.html.erb [hunk 44]
```diff
@@ -0,0 +1,66 @@
...
+                    
+                    <% else %>
+                      <td>Document</td>
+                    <% end %>
+                    <td><%= index_no + 1 %></td>
+                    <td><%=  document.document_file_name %></td>
+                    <td><%=  document.document_content_type %></td>
+                    <td><%= link_to "Open", document.document.url(:original, false) %></td>
+                  
+                    <td><%= link_to "Delete", issue_document_path(@issue_id ,document), method: :delete, data:{confirm: 'Are you sure?'}, :class => 'btn btn-primary' %></td>
```
Comment (mujtaba7vals) [inline app/views/documents/index.html.erb]: spacing

File: app/views/issues/show.html.erb [hunk 45]
```diff
@@ -40,6 +40,41 @@
...
+          <% end %>
+      </ul>
+    </div>
+  <% end %>
+
+
+  <div class="form-group">
+    <%= f.label :document %>
+    <%= f.file_field :document, class: 'form-control' %>
+
```
Comment (mujtaba7vals) [inline app/views/issues/show.html.erb]: spacing

File: app/views/projects/_form.html.erb [hunk 46]
```diff
@@ -0,0 +1,25 @@
+<%= form_for @project do |f| %>
+  <% if @project.errors.any? %>
+    <div id="error_explanation">
+      <h2><%= pluralize(@project.errors.count, "error") %> prohibited this project from being saved:</h2>
+
+      <ul>
+      <% @project.errors.full_messages.each do |message| %>
```
Comment (mujtaba7vals) [inline app/views/projects/_form.html.erb]: verify does it work

File: app/views/projects/_mediumprioritybtn.html.erb [hunk 47]
```diff
@@ -0,0 +1 @@
+<button type="button" class="btn btn-warning btn-sm">Medium</button>
```
Comment (mujtaba7vals) [inline app/views/projects/_mediumprioritybtn.html.erb]: mediumprioritybtn.html.erb not a good file name

File: app/views/projects/_newcomment.html.erb [hunk 48]
```diff
@@ -0,0 +1,10 @@
+<div class="modal-header">
+  <button type="button" class="close" data-dismiss="modal">x</button>
```
Comment (mujtaba7vals) [inline app/views/projects/_newcomment.html.erb]: need of this partial?

File: app/views/projects/_project.html.erb [hunk 49]
```diff
@@ -0,0 +1,16 @@
+
+<% if can?(:index, project) %>
+<div class="card">
```
Comment (mujtaba7vals) [inline app/views/projects/_project.html.erb]: indentation

File: app/views/projects/_project.html.erb [hunk 50]
```diff
@@ -0,0 +1,16 @@
...
+  <img class="card-img-top img-thumbnail" src="assets/default.jpg" alt="Card image cap">
+  <div class="card-body">
+    <h5 class="card-title"><%= link_to project.title, project%>
```
Comment (mujtaba7vals) [inline app/views/projects/_project.html.erb]: spacing

File: app/views/projects/show.html.erb [hunk 51]
```diff
@@ -0,0 +1,106 @@
...
+        <% elsif issue.priority == 1 %>
+          <td ><span class="glyphicon glyphicon-arrow-right green"></span></td>
+        <% elsif issue.priority == 0 %>
+          <td ><span class="glyphicon glyphicon-arrow-down blue"></span></td>
+        <% end %>
+
+        <td><%= issue.start_date %></td>
+
+        <td><%= issue.due_date %></td>
+      </tr>
```
Comment (mujtaba7vals) [inline app/views/projects/show.html.erb]: indentation

File: app/views/projects/show.html.erb [hunk 52]
```diff
@@ -0,0 +1,106 @@
...
+      <% end %>
+    </tbody>
+  </table>
+</div>
+
+<%= render 'comments/comment' %>
+
+<%= link_to 'Show all Documents', project_documents_path(@project),class: 'btn btn-primary' %>
```
Comment (mujtaba7vals) [inline app/views/projects/show.html.erb]: spacing

File: app/controllers/documents_controller.rb [hunk 53]
```diff
@@ -0,0 +1,93 @@
+class DocumentsController < ApplicationController
+  load_and_authorize_resource
```
Comment (Zeeshan-Siddhu) [inline app/controllers/documents_controller.rb]: load issue and project through cancan

File: app/controllers/teams_controller.rb [hunk 54]
```diff
@@ -0,0 +1,111 @@
...
+  def new
+    @team = Team.new
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def show
+    @team = Team.find(params[:id])
+    @team_memberships = TeamMembership.where('team_id = ? ', @team.id)
```
Comment (Zeeshan-Siddhu) [inline app/controllers/teams_controller.rb]: @team.team_memberships

File: app/controllers/teams_controller.rb [hunk 55]
```diff
@@ -0,0 +1,111 @@
...
+      format.html
+    end
+  end
+
+  def create
+    @team = Team.new(team_params)
+    @team.company_id = current_tenant.id ## need to be changed
+
+    if @team.save
+      TeamMembership.create!(is_team_admin: true, is_approved: false, team_id: Team.last.id, user_id: current_user.id)
```
Comment (Zeeshan-Siddhu) [inline app/controllers/teams_controller.rb]: @team.team_memberships.create

**Pull request closed without merging**